zebra-consensus/src/checkpoint.rs

```diff
@@ -0,0 +1,158 @@
+//! Checkpoint-based block verification for Zebra.
+//!
+//! Checkpoint-based verification uses a list of checkpoint hashes to speed up the
+//! initial chain sync for Zebra. This list is distributed with Zebra.
+//!
+//! The CheckpointVerifier compares each block's `BlockHeaderHash` against the known
+//! checkpoint hashes. If it matches, then the block is verified, and added to the
+//! `ZebraState`. Otherwise, if the block's height is lower than the maximum checkpoint
+//! height, the block awaits the verification of its child block.
+//!
+//! Verification is provided via a `tower::Service`, to support backpressure and batch
+//! verification.
+
+use futures_util::FutureExt;
+use std::{
+    collections::HashMap,
+    error,
+    future::Future,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+use tower::{buffer::Buffer, Service};
+
+use zebra_chain::block::{Block, BlockHeaderHash};
+use zebra_chain::types::BlockHeight;
+
+struct CheckpointVerifier<S> {
+    /// The underlying `ZebraState`.
+    state_service: S,
+
+    /// Each checkpoint consists of a coinbase height and block header hash.
+    ///
+    /// Checkpoints should be chosen to avoid forks or chain reorganizations,
+    /// which only happen in the last few hundred blocks in the chain.
+    /// (zcashd allows chain reorganizations up to 99 blocks, and prunes
+    /// orphaned side-chains after 288 blocks.)
+    checkpoint_list: Option<HashMap<BlockHeight, BlockHeaderHash>>,
+}
+
+/// The error type for the CheckpointVerifier Service.
+// TODO(jlusby): Error = Report ?
+type Error = Box<dyn error::Error + Send + Sync + 'static>;
+
+/// The CheckpointVerifier service implementation.
+///
+/// After verification, blocks are added to the underlying state service.
+impl<S> Service<Arc<Block>> for CheckpointVerifier<S>
+where
+    S: Service<zebra_state::Request, Response = zebra_state::Response, Error = Error>,
+    S::Future: Send + 'static,
+{
+    type Response = BlockHeaderHash;
+    type Error = Error;
+    type Future =
+        Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send + 'static>>;
+
+    fn poll_ready(&mut self, context: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.state_service.poll_ready(context)
+    }
+
+    fn call(&mut self, block: Arc<Block>) -> Self::Future {
+        // TODO(jlusby): Error = Report, handle errors from state_service.
+
+        // These checks are cheap, so we can do them in the call()
```

Comment:
I'm not sure if its worth keeping these out of the future, this handling code can be simplified a lot if its all inside of the final `async move` block